File: src/procclean/formatters/columns.py
"""Column specifications for process tables."""

from collections.abc import Callable
from dataclasses import dataclass

from procclean.core import ProcessInfo


def clip_right(s: str, max_len: int) -> str:
    """Truncate string from right, keeping left portion."""
    return s if len(s) <= max_len else s[:max_len]


def clip_left(s: str, max_len: int) -> str:
    """Truncate string from left, keeping right portion with ellipsis."""
    return s if len(s) <= max_len else "..." + s[-(max_len - 3) :]


def fmt_float1(v: object) -> str:
    """Format as float with 1 decimal."""
    return f"{float(v):.1f}"  # type: ignore[arg-type]


def fmt_status(p: ProcessInfo) -> str:
    """Format status with optional markers."""
    status = p.status
    if p.is_orphan:
        status += " [orphan]"
    if p.in_tmux:
        status += " [tmux]"
    return status


@dataclass(frozen=True)
class ColumnSpec:
    """Specification for a table column."""

    key: str
    header: str
    get: Callable[[ProcessInfo], object]
    fmt: Callable[[object], str] = str
    max_width: int | None = None

    def extract(self, proc: ProcessInfo) -> str:
        """Extract and format value from process."""
        raw = self.get(proc)
        formatted = self.fmt(raw)
        if self.max_width and len(formatted) > self.max_width:
            # Use clip_left for paths (cwd), clip_right for names
            if self.key == "cwd":
                return clip_left(formatted, self.max_width)
            return clip_right(formatted, self.max_width)
        return formatted


# Column definitions - single source of truth
COLUMNS: dict[str, ColumnSpec] = {
    "pid": ColumnSpec("pid", "PID", lambda p: p.pid),
    "name": ColumnSpec("name", "Name", lambda p: p.name, max_width=25),
    "rss_mb": ColumnSpec("rss_mb", "RAM (MB)", lambda p: p.rss_mb, fmt_float1),
    "cpu_percent": ColumnSpec(
        "cpu_percent", "CPU%", lambda p: p.cpu_percent, fmt_float1
    ),
    "cwd": ColumnSpec("cwd", "CWD", lambda p: p.cwd, max_width=40),
    "ppid": ColumnSpec("ppid", "PPID", lambda p: p.ppid),
    "parent_name": ColumnSpec(
        "parent_name", "Parent", lambda p: p.parent_name, max_width=15
    ),
    "status": ColumnSpec("status", "Status", fmt_status),
    "cmdline": ColumnSpec("cmdline", "Command", lambda p: p.cmdline, max_width=60),
    "username": ColumnSpec("username", "User", lambda p: p.username),
}

DEFAULT_COLUMNS: tuple[str, ...] = (
    "pid",
    "name",
    "rss_mb",
    "cpu_percent",
    "cwd",
    "ppid",
    "status",
)


def get_available_columns() -> list[str]:
    """Return list of available column keys."""
    return list(COLUMNS.keys())

File: src/formatters/columns.rs
use crate::core::models::ProcessInfo;

/// Side to clip when truncating text
#[derive(Debug, Clone, Copy)]
pub enum ClipSide {
    /// Keep the right portion (add "..." on left)
    Left,
    /// Keep the left portion (add "..." on right)
    Right,
}

/// Clip text to max length with ellipsis
pub fn clip(text: &str, max_len: usize, side: ClipSide) -> String {
    if text.len() <= max_len {
        return text.to_string();
    }

    match side {
        ClipSide::Left => {
            let keep = max_len.saturating_sub(3);
            format!("...{}", &text[text.len().saturating_sub(keep)..])
        }
        ClipSide::Right => {
            let keep = max_len.saturating_sub(3);
            format!("{}...", &text[..keep])
        }
    }
}

/// Column specification for formatting
pub struct ColumnSpec {
    pub key: &'static str,
    pub header: &'static str,
    pub max_width: Option<usize>,
    pub clip_side: ClipSide,
}

impl ColumnSpec {
    /// Extract and format the value for this column
    pub fn extract(&self, proc: &ProcessInfo) -> String {
        let value = match self.key {
            "pid" => proc.pid.to_string(),
            "name" => proc.name.clone(),
            "rss_mb" => format!("{:.1}", proc.rss_mb),
            "cpu_percent" => format!("{:.1}", proc.cpu_percent),
            "cwd" => proc.cwd.clone(),
            "ppid" => proc.ppid.to_string(),
            "parent_name" => proc.parent_name.clone(),
            "status" => proc.display_status(),
            "cmdline" => proc.cmdline.clone(),
            "username" => proc.username.clone(),
            _ => "?".to_string(),
        };

        if let Some(max_width) = self.max_width {
            clip(&value, max_width, self.clip_side)
        } else {
            value
        }
    }
}

/// Available columns
pub const COLUMNS: &[ColumnSpec] = &[
    ColumnSpec {
        key: "pid",
        header: "PID",
        max_width: None,
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "name",
        header: "Name",
        max_width: Some(25),
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "rss_mb",
        header: "RAM (MB)",
        max_width: None,
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "cpu_percent",
        header: "CPU%",
        max_width: None,
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "cwd",
        header: "CWD",
        max_width: Some(35),
        clip_side: ClipSide::Left,
    },
    ColumnSpec {
        key: "ppid",
        header: "PPID",
        max_width: None,
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "parent_name",
        header: "Parent",
        max_width: Some(20),
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "status",
        header: "Status",
        max_width: Some(40),
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "cmdline",
        header: "Command",
        max_width: Some(60),
        clip_side: ClipSide::Right,
    },
    ColumnSpec {
        key: "username",
        header: "User",
        max_width: Some(15),
        clip_side: ClipSide::Right,
    },
];

/// Default columns to display
pub const DEFAULT_COLUMN_KEYS: &[&str] = &["pid", "name", "rss_mb", "cpu_percent", "cwd", "ppid", "status"];

/// Get column specs by keys
pub fn get_columns(keys: &[&str]) -> Vec<&'static ColumnSpec> {
    keys.iter()
        .filter_map(|key| COLUMNS.iter().find(|col| col.key == *key))
        .collect()
}

/// Get default columns
pub fn get_default_columns() -> Vec<&'static ColumnSpec> {
    get_columns(DEFAULT_COLUMN_KEYS)
}
